benches/nexmark/main.rs

```diff
@@ -1 +1,67 @@
+//! Nexmark benchmarks for DBSP
+//!
+//! Based on the queries defined at
+#![feature(is_some_with)]
+use anyhow::Result;
+use clap::Parser;
+use dbsp::{
+    circuit::Root,
+    nexmark::{
+        config::Config as NexmarkConfig, generator::config::Config as GeneratorConfig,
+        model::Event, queries::q1, NexmarkSource,
+    },
+    profile::CPUProfiler,
+    trace::{ord::OrdZSet, BatchReader},
+};
+use rand::prelude::ThreadRng;
 
+// TODO: Enable running specific tests
+fn main() -> Result<()> {
+    let nexmark_config = NexmarkConfig::parse();
+    let generator_config = GeneratorConfig::new(nexmark_config, 0, 0, 0, 0);
+
+    let cpu_profiler = CPUProfiler::new();
+    let root = Root::build(|circuit| {
+        // TODO(absoludity): CPUProfiler is not currently used in any of the
+        // other benchmarks, only commented out. Not yet sure whether it will be
+        // helpful either. Some tools mentioned at
+        // https://nnethercote.github.io/perf-book/benchmarking.html but as had
+        // been said earlier, most are more suited to micro-benchmarking.  I
+        // assume that our best option for comparable benchmarks will be to try
+        // to do exactly what the Java implementation does: core(s) * time [see
+        // Run Nexmark](https://github.com/nexmark/nexmark#run-nexmark).  Right
+        // now, just grab simple timestamp to do a duration for 100_000_000 events with
+        // a source generating 1M events/sec as per the Nexmark results.
+        let start_time = std::time::SystemTime::now();
+        cpu_profiler.attach(&circuit, "cpu profiler");
+
+        let source =
+            NexmarkSource::<ThreadRng, isize, OrdZSet<Event, isize>>::new(generator_config);
+        let input = circuit.add_source(source);
+
+        let output = q1(input);
+        let mut events_processed: u64 = 0;
+
+        output.inspect(move |zs: &OrdZSet<_, _>| {
+            events_processed += zs.len() as u64;
+            if events_processed >= 100_000_000 {
+                println!(
+                    "{} events processed in {}ms",
+                    events_processed,
+                    std::time::SystemTime::now()
+                        .duration_since(start_time)
+                        .unwrap()
+                        .as_millis()
+                );
+                // TODO: send signal through channel to stop iterating once
+                // sufficient data for test is processed?
+            }
+        });
+    })
+    .unwrap();
+
+    for _ in 0..100_000 {
```

Comment:
I'll work on a circuit lifecycle API, so we won't need to hardcode fixed numbers of iterations.

Reply:
Great, thanks. Noticed that sometimes, given the way we're batching whatever is ready, this estimate works if CPU keeps up, but as soon as the CPU is busy with other things and a backlog generates, it seems to cascade (as the batches (and mem?) grow massively)

Reply:
Hey @ryzhyk , do you know any way I can work around this until the circuit lifecycle API is available? Specifically, AFAICT, I either need to be able to:
1. communicate from the closure within the circuit builder (from `output.inspect()`) that the desired number of events have been processed, so that I can stop calling `root.step()`, or
2. have the circuit notice that the source has reached a fixed point and the output is no longer changing (hmm, though maybe this can't be used given delay operators etc., not sure).

As per my comment, I was initially thinking I'd use a channel as a workaround to communicate out (given any var captured by the circuit builder closure as a `&mut` won't be accessible after the circuit is built), but I hadn't realised it's not actually async (as in, I'm manually calling `root.step()` rather than a scheduler in another thread or similar.

No worries if you don't have time, just in case you've something obvious I've missed. Thanks.

Reply:
I think you can just create a variable of type `Rc<Cell<bool>>` and `set` it inside the source operator when all outputs have been generated. You can check the value of the variable inside the `step()` loop, aborting on `true`.

And yes, this is something that the I/O API will solve, for both single-threaded and multi-threaded cases (i.e., when using `Runtime`).

Reply:
Thanks @ryzhyk, yep, that works fine. Now I'm needing to create a macro that generates the circuits for each query it seems (given the queries have different return types). Anyway, will have that ready tomorrow I hope.